airflow/models/taskinstance.py

```diff
@@ -1237,25 +1293,32 @@ def check_and_change_state_before_execution(
         :param mark_success: Don't run the task, mark its state as success
         :param test_mode: Doesn't record success or failure in the DB
         :param job_id: Job (BackfillJob / LocalTaskJob / SchedulerJob) ID
-        :param pool: specifies the pool to use to run the task instance
+        :param pool: Specifies the pool to use to run the task instance
         :param external_executor_id: The identifier of the celery executor
         :param session: SQLAlchemy ORM Session
         :return: whether the state was changed to running or not
         """
-        task = self.task
-        self.refresh_from_task(task, pool_override=pool)
-        self.test_mode = test_mode
-        self.refresh_from_db(session=session, lock_for_update=True)
-        self.job_id = job_id
-        self.hostname = get_hostname()
-        self.pid = None
+        ti = TaskInstance.retrieve_from_db(
+            dag_id, run_id, task_id, map_index, session=session, lock_for_update=True
+        )
+        if ti is None:
```

Comment:
Sometime this function is called but there is not yet a db record for the task instance. cf `test_check_and_change_state_before_execution_dep_not_met` which highlight this case.